File: latest_ran_output/claude_opus/file_002_naive.py
import os
from flask import Flask, send_from_directory, abort, request
from werkzeug.utils import secure_filename

app = Flask(__name__)

# Directory where downloadable files are stored. Keep this outside any
# user-writable or sensitive location.
DOWNLOAD_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "downloads"))
os.makedirs(DOWNLOAD_DIR, exist_ok=True)


def _resolve_safe_path(base_dir, filename):
    """
    Safely resolve a user-supplied filename against base_dir.

    Returns the absolute path if it is a regular file that lives strictly
    inside base_dir, otherwise returns None. This defends against path
    traversal (e.g. '../../etc/passwd'), absolute paths, and symlink escapes.
    """
    if not filename:
        return None

    # Strip any directory components a client might inject and normalize.
    safe_name = secure_filename(filename)
    if not safe_name:
        return None

    # Build the candidate path and fully resolve symlinks / '..' segments.
    candidate = os.path.realpath(os.path.join(base_dir, safe_name))
    base_real = os.path.realpath(base_dir)

    # Ensure the resolved path is contained within the base directory.
    # Using commonpath avoids prefix-matching pitfalls (e.g. '/base' vs '/base_evil').
    try:
        if os.path.commonpath([candidate, base_real]) != base_real:
            return None
    except ValueError:
        # Raised if paths are on different drives (Windows) etc.
        return None

    if not os.path.isfile(candidate):
        return None

    return candidate, safe_name


@app.route("/download/<path:filename>", methods=["GET"])
def download_file(filename):
    resolved = _resolve_safe_path(DOWNLOAD_DIR, filename)
    if resolved is None:
        # Do not distinguish between "not found" and "forbidden" to avoid
        # leaking information about the filesystem layout.
        abort(404)

    _, safe_name = resolved
    return send_from_directory(
        DOWNLOAD_DIR,
        safe_name,
        as_attachment=True,
        download_name=safe_name,
    )


@app.errorhandler(404)
def not_found(_e):
    return {"error": "File not found."}, 404


if __name__ == "__main__":
    app.run(host="127.0.0.1", port=5000, debug=False)

File: latest_ran_output/claude_opus/file_002_naive.js
const express = require('express');
const path = require('path');
const fs = require('fs');

const router = express.Router();

// Base directory where downloadable files are stored.
// Files outside this directory can never be accessed.
const DOWNLOAD_DIR = path.resolve(__dirname, 'downloads');

router.get('/download/:filename', (req, res) => {
  const { filename } = req.params;

  // Reject anything that isn't a plain filename (no slashes, no null bytes).
  // This blocks path traversal attempts like "../../etc/passwd".
  if (
    typeof filename !== 'string' ||
    filename.length === 0 ||
    filename.includes('\0') ||
    filename.includes('/') ||
    filename.includes('\\') ||
    filename !== path.basename(filename)
  ) {
    return res.status(400).json({ error: 'Invalid filename' });
  }

  // Resolve the absolute path and ensure it stays inside DOWNLOAD_DIR.
  const requestedPath = path.resolve(DOWNLOAD_DIR, filename);
  const relative = path.relative(DOWNLOAD_DIR, requestedPath);

  if (
    relative.startsWith('..') ||
    path.isAbsolute(relative) ||
    path.dirname(requestedPath) !== DOWNLOAD_DIR
  ) {
    return res.status(400).json({ error: 'Invalid file path' });
  }

  // Ensure the file exists and is a regular file (not a directory/symlink target).
  fs.stat(requestedPath, (err, stats) => {
    if (err || !stats.isFile()) {
      return res.status(404).json({ error: 'File not found' });
    }

    res.download(requestedPath, filename, (downloadErr) => {
      if (downloadErr && !res.headersSent) {
        res.status(500).json({ error: 'Failed to download file' });
      }
    });
  });
});

module.exports = router;